File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def can_change(word_A, word_B):
    n = len(word_A)
    different_cnt = 0
    
    for i in range(n):
        if word_A[i] == word_B[i]:
            continue
        else:
            different_cnt += 1
    
    if different_cnt == 1:
        return True
    else:
        return False

def BFS(begin, target, words, visited, n):
    q = deque([])
    q.append([begin, visited, 0])
    
    while q:
        now_word, now_visited, now_depth = q.popleft()
        
        if now_word == target:
            return now_depth
        
        for i in range(n):
            if now_visited[i] == False:
                next_word = words[i]
                if can_change(now_word, next_word) == True:
                    next_visited = now_visited[:]
                    next_visited[i] = True
                    q.append([next_word, next_visited, now_depth + 1])
            
    return 0
            
    

def solution(begin, target, words):
    n = len(words)
    visited = [False] * n
    
    answer = BFS(begin, target, words, visited, n)
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
// 너비우선 BFS 문제!

import java.util.*;

class Solution {
    boolean[] visited;
    
    public boolean check(String before, String after) {
        boolean can_convert = false;
        int n = before.length();
        int different_cnt = 0;  // 알파벳 다른거 세기
        
        for (int i = 0; i < n; i++) {
            if (different_cnt >= 2) {
                break;
            }
            
            if (before.charAt(i) != after.charAt(i)) {
                different_cnt += 1;
            }
        }
        
        if (different_cnt == 1) {
            can_convert = true;
        }
        
        return can_convert;
    }
    
    public int BFS(String begin, String target, String[] words) {
        int result = 0;
        Queue<Object[]> q = new LinkedList<>();
        q.offer(new Object[]{begin, 0});

        while (q.size() != 0) {
            Object[] now = q.poll();
            String now_word = (String) now[0];
            int now_depth = (int) now[1];
            
            // 타겟을 찾으면
            if (now_word.equals(target)) {
                result = now_depth;
                break;
            }
            
            for (int i = 0; i < words.length; i++) {
                if (visited[i] == false) {
                    boolean can_convert = check(now_word, words[i]);
                    
                    if (can_convert == true) {
                        q.offer(new Object[] {words[i], now_depth + 1});
                        visited[i] = true;
                    }
                }
            }
            
        }
        
        return result;
    }
    
    public int solution(String begin, String target, String[] words) {
        int n = words.length;
        visited = new boolean[n];
        
        for(int i = 0; i < n; i++) {
            visited[i] = false;
        }
        
        
        int answer = BFS(begin, target, words);
        return answer;
    }
}